ee/desktop/runner/runner.go

```diff
@@ -0,0 +1,426 @@
+	process    *os.Process
+	path       string
+	socketPath string
+}
+
+// New creates and returns a new DesktopUsersProcessesRunner runner and initializes all required fields
+func New(opts ...DesktopUsersProcessesRunnerOption) *DesktopUsersProcessesRunner {
+	runner := &DesktopUsersProcessesRunner{
+		logger:           log.NewNopLogger(),
+		interrupt:        make(chan struct{}),
+		uidProcs:         make(map[string]processRecord),
+		updateInterval:   time.Second * 5,
+		procsWg:          &sync.WaitGroup{},
+		interruptTimeout: time.Second * 10,
+		usersFilesRoot:   filepath.Join(os.TempDir(), "kolide-desktop"),
+	}
+
+	for _, opt := range opts {
+		opt(runner)
+	}
+
+	return runner
+}
+
+// Execute immediately checks if the current console user has a desktop process running. If not, it will start a new one.
+// Then repeats based on the executionInterval.
+func (r *DesktopUsersProcessesRunner) Execute() error {
+	f := func() {
+		if err := r.runConsoleUserDesktop(); err != nil {
+			level.Error(r.logger).Log("msg", "running console user desktop", "err", err)
+		}
+	}
+
+	f()
+
+	ticker := time.NewTicker(r.updateInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			f()
+		case <-r.interrupt:
+			level.Debug(r.logger).Log("msg", "interrupt received, exiting desktop execute loop")
+			return nil
+		}
+	}
+}
+
+// Interrupt stops creating launcher desktop processes and kills any existing ones.
+func (r *DesktopUsersProcessesRunner) Interrupt(interruptError error) {
+	level.Debug(r.logger).Log(
+		"msg", "sending interrupt to desktop users processes runner",
+		"err", interruptError,
+	)
+
+	r.interrupt <- struct{}{}
+
+	wgDone := make(chan struct{})
+	go func() {
+		defer close(wgDone)
+		r.procsWg.Wait()
+	}()
+
+	for uid, proc := range r.uidProcs {
+		client := client.New(r.authToken, proc.socketPath)
+		if err := client.Shutdown(); err != nil {
+			level.Error(r.logger).Log(
+				"msg", "error sending shutdown command to desktop process",
+				"uid", uid,
+				"pid", proc.process.Pid,
+				"path", proc.path,
+				"err", err,
+			)
+		}
+	}
+
+	select {
+	case <-wgDone:
+		level.Debug(r.logger).Log("msg", "all desktop processes shutdown successfully")
+		return
+	case <-time.After(r.interruptTimeout):
+		level.Error(r.logger).Log("msg", "timeout waiting for desktop processes to exit, now killing")
+		for uid, processRecord := range r.uidProcs {
+			if !processExists(processRecord) {
+				continue
+			}
+			if err := processRecord.process.Kill(); err != nil {
+				level.Error(r.logger).Log(
+					"msg", "error killing desktop process",
+					"uid", uid,
+					"pid", processRecord.process.Pid,
+					"path", processRecord.path,
+					"err", err,
+				)
+			}
+		}
+	}
+}
+
+func (r *DesktopUsersProcessesRunner) runConsoleUserDesktop() error {
+	executablePath, err := r.determineExecutablePath()
+	if err != nil {
+		return fmt.Errorf("determining executable path: %w", err)
+	}
+
+	consolerUsers, err := r.consoleUsers()
```

Comment:
Is the`r` a typo?

Reply:
not a typo, but there is no reason to hang consoleUsers() on the runner struct, it's now un hanged

Reply:
Heh. I didn't mean `r.` I mean `consolerUsers`. What's with the r in the middle?

Reply:
oops, fixed =)

Reply:
I keep wanting it to be like `sommelier`, someone who can show me exquisitely crafted consoles from around the world. 🤣